### PR TITLE
Rewrite cache_data to 4.x API

### PR DESCRIPTION
**Pull request by ekohl**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,8 +3,7 @@ Gemfile.lock
 Gemfile.local
 vendor/
 .vendor/
-spec/fixtures/manifests/
-spec/fixtures/modules/
```
Comment (bastelfreak) [inline .gitignore]: This comes from modulesync_config and this is on purpose. Some modules ship files as fixtures. For example https://github.com/voxpupuli/puppet-prometheus/tree/master/spec/fixtures/files

File: lib/puppet/functions/cache_data.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
+require 'fileutils'
+require 'yaml'
+require 'etc'
```
Comment (bastelfreak) [inline lib/puppet/functions/cache_data.rb]: where is that coming from? I assume https://rubygems.org/gems/etc ?

Comment (ekohl) [reply lib/puppet/functions/cache_data.rb]: It was already there. It's https://ruby-doc.org/stdlib-2.5.1/libdoc/etc/rdoc/Etc.html

File: .gitignore (hunk 1 above)
Comment (ekohl) [reply .gitignore]: This now creates the tmpdir inside the fixtures so it needs a change for that at least.

File: lib/puppet/functions/cache_data.rb (hunk 2 above)
Comment (bastelfreak) [reply lib/puppet/functions/cache_data.rb]: ah nice.

**Comment by alexjfisher** [comment]
@ekohl Would you like to cover this off https://github.com/voxpupuli/puppet-extlib/issues/80#issuecomment-352534611 in this PR?

**Comment by alexjfisher** [comment]
Looks like we're not getting a good result in the table of contents.
https://github.com/voxpupuli/puppet-extlib/pull/114/files#diff-eb120e74eba90da6132666ceb0405bf9L8

**Comment by ekohl** [comment]
Updated

File: lib/puppet/functions/cache_data.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+# @summary Retrieves data from a cache file, or creates it with supplied data if the file doesn't exist
```
Comment (alexjfisher) [inline lib/puppet/functions/cache_data.rb]: ooh, so `@summary` did work?  The puppet-strings style guide doesn't say to use it.
https://tickets.puppetlabs.com/browse/PDOC-263 also mentions this, but says it does fix the TOC.

File: REFERENCE.md [hunk 4]
```diff
@@ -19,27 +19,39 @@
 
 ### cache_data
 
-Type: Ruby 3.x API
+Type: Ruby 4.x API
 
-Retrieves data from a cache file, or creates it with supplied data if the file doesn't exist
```
Comment (alexjfisher) [inline REFERENCE.md]: Is it a shame the summary line doesn't appear here too?

**Comment by ekohl** [comment]
Updated to repeat the summary since I agree it's more readable as a user.

**Pull request merged**